File: ConsoleApp/src/main/java/MindView/SimpleMicroBanchmark.java
package MindView;

import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

abstract class Incrementable {
    protected long counter = 0;

    public abstract void increment();
}

class SynchronizingTest extends Incrementable {
    @Override
    public synchronized void increment() {
        ++counter;
    }
}

class LockingTest extends Incrementable {
    private Lock lock = new ReentrantLock();

    @Override
    public void increment() {
        lock.lock();
        try {
            ++counter;
        } finally {
            lock.unlock();
        }
    }
}

public class SimpleMicroBanchmark {
    static long test(Incrementable incrementable) {
        long start = System.nanoTime();
        for (long i = 0; i < 10000000L; i++) {
            incrementable.increment();
        }
        return System.nanoTime() - start;
    }

    public static void main(String[] args) {
        long syncTime = test(new SynchronizingTest());
        long lockTime = test(new LockingTest());
        System.out.printf("synchronized: %1$10d\n", syncTime);
        System.out.printf("Lock:         %1$10d\n", lockTime);
        System.out.printf("Lock/synchronized = %1$.3f", (double) lockTime / (double) syncTime);
    }
}

File: ConsoleApp/src/main/kotlin/consoleapp/mindview/SimpleMicroBanchmark.kt
package consoleapp.mindview

import java.util.concurrent.locks.ReentrantLock

internal abstract class Incrementable {
    protected var counter: Long = 0

    abstract fun increment()
}

internal class SynchronizingTest : Incrementable() {
    @Synchronized override fun increment() {
        ++counter
    }
}

internal class LockingTest : Incrementable() {
    private val lock = ReentrantLock()

    override fun increment() {
        lock.lock()
        try {
            ++counter
        } finally {
            lock.unlock()
        }
    }
}

object SimpleMicroBanchmark {
    internal fun test(incrementable: Incrementable): Long {
        val start = System.nanoTime()
        for (i in 0..10000000L - 1) {
            incrementable.increment()
        }
        return System.nanoTime() - start
    }

    @JvmStatic
    fun main(args: Array<String>) {
        val syncTime = test(SynchronizingTest())
        val lockTime = test(LockingTest())
        System.out.printf("synchronized: %1$10d\n", syncTime)
        System.out.printf("Lock:         %1$10d\n", lockTime)
        System.out.printf("Lock/synchronized = %1$.3f", lockTime.toDouble() / syncTime.toDouble())
    }
}
